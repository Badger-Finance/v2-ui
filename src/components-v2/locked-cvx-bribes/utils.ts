import BigNumber from 'bignumber.js';
import { formatWithoutExtraZeros, numberWithCommas, unscale } from '../../mobx/utils/helpers';

export function formatBalance(balance: BigNumber): string {
  return numberWithCommas(formatWithoutExtraZeros(unscale(balance, 18)));
}
