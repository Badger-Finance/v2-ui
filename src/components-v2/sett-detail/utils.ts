import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import { ChartMode } from '../../mobx/model/setts/sett-charts';
import { DelaySeverity } from '../../mobx/model/setts/sett-rewards';
import { SettDetailMode } from '../../mobx/model/setts/sett-detail';

dayjs.extend(utc);

export const ChartModeTitles = {
	[ChartMode.value]: 'Sett Value',
	[ChartMode.ratio]: 'Token Ratio',
	[ChartMode.accountBalance]: 'Your Holdings',
};

export const SettModeTitles = {
	[SettDetailMode.settInformation]: 'Vault Info',
	[SettDetailMode.userInformation]: 'User Info',
};

export const calculateDelaySeverity = (delay: number): DelaySeverity => {
	if (delay >= 4) {
		return DelaySeverity.high;
	}

	if (delay >= 2) {
		return DelaySeverity.medium;
	}

	return DelaySeverity.none;
};

export const calculateDifferenceInHoursFromCycle = (cycle: Date): number => {
	return Math.abs(dayjs(cycle).diff(dayjs(), 'hours'));
};
