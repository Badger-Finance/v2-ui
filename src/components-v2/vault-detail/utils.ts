import { Vault, VaultData } from '@badger-dao/sdk';
import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import { ChartMode } from '../../mobx/model/vaults/vault-charts';
import { DelaySeverity } from '../../mobx/model/vaults/vault-rewards';

dayjs.extend(utc);

export const ChartModeTitles: Record<string, string> = {
	[ChartMode.Value]: 'Sett Value',
	[ChartMode.Ratio]: 'Token Ratio',
	[ChartMode.AccountBalance]: 'My Holdings',
	[ChartMode.BoostMultiplier]: 'Badger Boost',
};

export const calculateDelaySeverity = (delay: number): DelaySeverity => {
	if (delay >= 4) {
		return DelaySeverity.high;
	}

	if (delay >= 2) {
		return DelaySeverity.medium;
	}

	return DelaySeverity.none;
};

export const calculateDifferenceInHoursFromCycle = (cycle: Date): number => {
	return Math.abs(dayjs(cycle).diff(dayjs(), 'hours'));
};

export function defaultVaultBalance(vault: Vault): VaultData {
	return {
		address: vault.vaultToken,
		name: vault.name,
		symbol: vault.asset,
		pricePerFullShare: vault.pricePerFullShare,
		balance: 0,
		value: 0,
		earnedBalance: 0,
		earnedValue: 0,
		depositedBalance: 0,
		withdrawnBalance: 0,
		tokens: [],
		earnedTokens: [],
	};
}

export interface Balance {
	balance: number;
}

export function hasBalance(input?: Balance): boolean {
	if (!input) {
		return false;
	}
	return input.balance > 0;
}
