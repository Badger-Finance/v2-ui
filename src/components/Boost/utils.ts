import { makeStyles } from '@material-ui/core/styles';
import { ClassNameMap } from '@material-ui/core/styles/withStyles';
import React from 'react';

type ComparisonConfig = {
  toCompareValue: number;
  toBeComparedValue: number;
  greaterCaseColor: string;
  lessCaseColor: string;
  defaultColor: string;
};

export const getColorFromComparison = ({
  toCompareValue,
  toBeComparedValue,
  greaterCaseColor,
  lessCaseColor,
  defaultColor,
}: ComparisonConfig): string => {
  if (toCompareValue > toBeComparedValue) return greaterCaseColor;
  if (toCompareValue < toBeComparedValue) return lessCaseColor;

  return defaultColor;
};

export const useAssetInputStyles = (
  currentValue: string,
  toCompare = 0,
): ((props?: React.CSSProperties) => ClassNameMap<'assetColor'>) => {
  return makeStyles((theme) => {
    const defaultColor = currentValue ? theme.palette.text.primary : theme.palette.text.secondary;
    const fontColor = getColorFromComparison({
      toCompareValue: Number(currentValue),
      toBeComparedValue: toCompare,
      greaterCaseColor: '#74D189',
      lessCaseColor: theme.palette.error.main,
      defaultColor,
    });

    return {
      assetColor: {
        color: fontColor,
      },
    };
  });
};
