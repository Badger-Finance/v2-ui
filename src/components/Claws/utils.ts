import BigNumber from 'bignumber.js';
import { INVALID_REASON } from './index';

export interface ValidationParams {
	amount: string | BigNumber;
	maximum: BigNumber;
	minimum?: BigNumber;
}

export function validateAmountBoundaries({
	amount,
	maximum,
	minimum = new BigNumber('0'),
}: ValidationParams): INVALID_REASON | undefined {
	let error: INVALID_REASON | undefined = undefined;

	const input = new BigNumber(amount);
	const amountExceedsBalance = input.gt(maximum);
	const amountIsLessThanMinimum = input.lt(minimum);

	if (amountIsLessThanMinimum) error = INVALID_REASON.UNDER_MINIMUM;
	if (amountExceedsBalance) error = INVALID_REASON.OVER_MAXIMUM;

	console.log({
		input: input.toString(),
		minimum: minimum.toString(),
		maximum: maximum.toString(),
		amountExceedsBalance,
		amountIsLessThanMinimum,
		error,
	});

	return error;
}
