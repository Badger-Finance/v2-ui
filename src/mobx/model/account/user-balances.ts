import { TokenBalance } from '../tokens/token-balance';

export interface TokenBalances {
	[contract: string]: TokenBalance;
}

export interface BalancesRequestAddresses {
	tokenAddresses: string[];
	generalSettAddresses: string[];
	guardedSettAddresses: string[];
	geyserAddresses: string[];
	userAddress: string;
	nonSettTokenAddresses?: string[];
}

export interface ExtractedBalances {
	tokenBalances: TokenBalances;
	settBalances: TokenBalances;
}

export interface GuestListInformation {
	guestLists: string[];
	guestListLookup: Record<string, string>;
}
