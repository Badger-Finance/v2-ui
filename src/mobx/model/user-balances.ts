import { TokenBalance } from './token-balance';

export interface UserBalances {
	[contract: string]: TokenBalance;
}
