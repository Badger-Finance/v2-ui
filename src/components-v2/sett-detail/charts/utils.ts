import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import { Sett } from '../../../mobx/model/setts/sett';
import { fetchSettChartInformation } from '../../../mobx/utils/apiV2';
import { SettSnapshotGranularity } from '../../../mobx/model/setts/sett-snapshot';
import { SettChartData } from '../../../mobx/model/setts/sett-charts';

dayjs.extend(utc);

export enum SettChartTimeframe {
	'day' = 1,
	'week' = 7,
	'month' = 30,
}

/**
 * Fetches chart information since the provided timeframe until the current date
 * @param sett
 * @param timeframe
 */
export const fetchSettChart = async (sett: Sett, timeframe: SettChartTimeframe): Promise<SettChartData[] | null> => {
	const timeframeDays = 1 * timeframe;
	const isDayTimeFrame = timeframe === SettChartTimeframe.day;

	// if timeframe is just one day then we want the granularity to be hours
	const granularity = isDayTimeFrame ? SettSnapshotGranularity.HOUR : SettSnapshotGranularity.DAY;

	const now = dayjs().utc(); // query until current date
	const from = dayjs(now).subtract(timeframeDays, 'days').utc();

	const fetchedData = await fetchSettChartInformation(sett.vaultToken, from.toDate(), now.toDate(), granularity);

	if (!fetchedData) {
		return null;
	}

	// data needs to be ascending sorted
	return fetchedData
		.sort((a, b) => a.timestamp - b.timestamp)
		.map((d) => ({ ...d, timestamp: new Date(d.timestamp) }));
};
