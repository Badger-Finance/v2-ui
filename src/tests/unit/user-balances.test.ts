import { extractBalanceRequestResults } from '../../mobx/utils/user-balances';
import { ContractCallResults } from 'ethereum-multicall';
import { BalanceNamespace } from '../../web3/config/namespaces';

const mockedCorrectResults: ContractCallResults = {
  results: {
    '0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D2B': {
      originalContractCallContext: {
        reference: '0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D2B',
        contractAddress: '0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D2B',
        abi: [], //empty on purpose, we're not using the abi and they take lots of space
        calls: [{ reference: 'balanceOf', methodName: 'balanceOf', methodParameters: ['0x00'] }],
        context: {
          namespace: BalanceNamespace.Token,
        },
      },
      callsReturnContext: [
        {
          returnValues: [{ amounts: '0x00' }],
          decoded: true,
          reference: 'balanceOf',
          methodName: 'balanceOf',
          methodParameters: ['0x00'],
          success: true,
        },
      ],
    },
    '0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D3B': {
      originalContractCallContext: {
        reference: '0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D3B',
        contractAddress: '0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D2B',
        abi: [], //empty on purpose, we're not using the abi and they take lots of space
        calls: [{ reference: 'balanceOf', methodName: 'balanceOf', methodParameters: ['0x00'] }],
        context: {
          namespace: BalanceNamespace.Sett,
        },
      },
      callsReturnContext: [
        {
          returnValues: [{ amounts: '0x00' }],
          decoded: true,
          reference: 'balanceOf',
          methodName: 'balanceOf',
          methodParameters: ['0x00'],
          success: true,
        },
      ],
    },
    '0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D4B': {
      originalContractCallContext: {
        reference: '0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D4B',
        contractAddress: '0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D4B',
        abi: [], //empty on purpose, we're not using the abi and they take lots of space
        calls: [{ reference: 'balanceOf', methodName: 'balanceOf', methodParameters: ['0x00'] }],
        context: {
          namespace: BalanceNamespace.GuardedSett,
        },
      },
      callsReturnContext: [
        {
          returnValues: [{ amounts: '0x00' }],
          decoded: true,
          reference: 'balanceOf',
          methodName: 'balanceOf',
          methodParameters: ['0x00'],
          success: true,
        },
      ],
    },
    '0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D5B': {
      originalContractCallContext: {
        reference: '0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D5B',
        contractAddress: '0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D5B',
        abi: [], //empty on purpose, we're not using the abi and they take lots of space
        calls: [{ reference: 'balanceOf', methodName: 'balanceOf', methodParameters: ['0x00'] }],
        context: {
          namespace: BalanceNamespace.Deprecated,
        },
      },
      callsReturnContext: [
        {
          returnValues: [{ amounts: '0x00' }],
          decoded: true,
          reference: 'balanceOf',
          methodName: 'balanceOf',
          methodParameters: ['0x00'],
          success: true,
        },
      ],
    },
    '0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D6B': {
      originalContractCallContext: {
        reference: '0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D6B',
        contractAddress: '0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D6B',
        abi: [], //empty on purpose, we're not using the abi and they take lots of space
        calls: [{ reference: 'balanceOf', methodName: 'balanceOf', methodParameters: ['0x00'] }],
        context: {
          namespace: BalanceNamespace.Deprecated,
        },
      },
      callsReturnContext: [
        {
          returnValues: [{ amounts: '0x00' }],
          decoded: true,
          reference: 'balanceOf',
          methodName: 'balanceOf',
          methodParameters: ['0x00'],
          success: true,
        },
      ],
    },
  },
  blockNumber: 0,
};

const mockedBadResults: ContractCallResults = {
  results: {
    '0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D2B': {
      originalContractCallContext: {
        reference: '0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D2B',
        contractAddress: '0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D2B',
        abi: [], //empty on purpose, we're not using the abi and they take lots of space
        calls: [{ reference: 'balanceOf', methodName: 'balanceOf', methodParameters: ['0x00'] }],
        context: {
          namespace: 'foo',
        },
      },
      callsReturnContext: [
        {
          returnValues: [{ amounts: '0x00' }],
          decoded: true,
          reference: '0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D2B',
          methodName: 'foo',
          methodParameters: [42],
          success: true,
        },
      ],
    },
  },
  blockNumber: 0,
};

describe('User balances utils', () => {
  describe('extractBalanceRequestResults', () => {
    it('extract results correctly', () => {
      const keys = extractBalanceRequestResults(mockedCorrectResults);
      expect(keys).toMatchSnapshot();
    });

    it('error logs invalid namespaces', () => {
      const mockedError = jest.fn();
      console.error = mockedError;
      extractBalanceRequestResults(mockedBadResults);
      expect(mockedError).toHaveBeenCalled();
    });
  });
});
