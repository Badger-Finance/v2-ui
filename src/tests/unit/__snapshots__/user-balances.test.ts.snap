// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User balances utils extractBalanceRequestResults extract results correctly 1`] = `
Object {
  "userDeprecatedSetts": Array [
    Object {
      "callsReturnContext": Array [
        Object {
          "decoded": true,
          "methodName": "balanceOf",
          "methodParameters": Array [
            "0x00",
          ],
          "reference": "balanceOf",
          "returnValues": Array [
            Object {
              "amounts": "0x00",
            },
          ],
          "success": true,
        },
      ],
      "originalContractCallContext": Object {
        "abi": Array [],
        "calls": Array [
          Object {
            "methodName": "balanceOf",
            "methodParameters": Array [
              "0x00",
            ],
            "reference": "balanceOf",
          },
        ],
        "context": Object {
          "namespace": "deprecated",
        },
        "contractAddress": "0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D5B",
        "reference": "0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D5B",
      },
    },
    Object {
      "callsReturnContext": Array [
        Object {
          "decoded": true,
          "methodName": "balanceOf",
          "methodParameters": Array [
            "0x00",
          ],
          "reference": "balanceOf",
          "returnValues": Array [
            Object {
              "amounts": "0x00",
            },
          ],
          "success": true,
        },
      ],
      "originalContractCallContext": Object {
        "abi": Array [],
        "calls": Array [
          Object {
            "methodName": "balanceOf",
            "methodParameters": Array [
              "0x00",
            ],
            "reference": "balanceOf",
          },
        ],
        "context": Object {
          "namespace": "deprecated",
        },
        "contractAddress": "0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D6B",
        "reference": "0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D6B",
      },
    },
  ],
  "userGeneralSetts": Array [
    Object {
      "callsReturnContext": Array [
        Object {
          "decoded": true,
          "methodName": "balanceOf",
          "methodParameters": Array [
            "0x00",
          ],
          "reference": "balanceOf",
          "returnValues": Array [
            Object {
              "amounts": "0x00",
            },
          ],
          "success": true,
        },
      ],
      "originalContractCallContext": Object {
        "abi": Array [],
        "calls": Array [
          Object {
            "methodName": "balanceOf",
            "methodParameters": Array [
              "0x00",
            ],
            "reference": "balanceOf",
          },
        ],
        "context": Object {
          "namespace": "sett",
        },
        "contractAddress": "0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D2B",
        "reference": "0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D3B",
      },
    },
  ],
  "userGuardedSetts": Array [
    Object {
      "callsReturnContext": Array [
        Object {
          "decoded": true,
          "methodName": "balanceOf",
          "methodParameters": Array [
            "0x00",
          ],
          "reference": "balanceOf",
          "returnValues": Array [
            Object {
              "amounts": "0x00",
            },
          ],
          "success": true,
        },
      ],
      "originalContractCallContext": Object {
        "abi": Array [],
        "calls": Array [
          Object {
            "methodName": "balanceOf",
            "methodParameters": Array [
              "0x00",
            ],
            "reference": "balanceOf",
          },
        ],
        "context": Object {
          "namespace": "guarded",
        },
        "contractAddress": "0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D4B",
        "reference": "0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D4B",
      },
    },
  ],
  "userTokens": Array [
    Object {
      "callsReturnContext": Array [
        Object {
          "decoded": true,
          "methodName": "balanceOf",
          "methodParameters": Array [
            "0x00",
          ],
          "reference": "balanceOf",
          "returnValues": Array [
            Object {
              "amounts": "0x00",
            },
          ],
          "success": true,
        },
      ],
      "originalContractCallContext": Object {
        "abi": Array [],
        "calls": Array [
          Object {
            "methodName": "balanceOf",
            "methodParameters": Array [
              "0x00",
            ],
            "reference": "balanceOf",
          },
        ],
        "context": Object {
          "namespace": "token",
        },
        "contractAddress": "0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D2B",
        "reference": "0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D2B",
      },
    },
  ],
}
`;
